csharp/src/Google.Protobuf/CodedOutputWriter.cs

```diff
@@ -0,0 +1,589 @@
+        public void WriteBool(bool value)
+        {
+            WriteRawByte(value ? (byte) 1 : (byte) 0);
+        }
+
+        /// <summary>
+        /// Writes a string field value, without a tag, to the stream.
+        /// The data is length-prefixed.
+        /// </summary>
+        /// <param name="value">The value to write</param>
+        public void WriteString(string value)
+        {
+            int length = Utf8Encoding.GetByteCount(value);
+            WriteLength(length);
+
+            Span<byte> buffer = writer.GetSpan(length);
+
+            if (buffer.Length >= length)
+            {
+                // Can write string to a single buffer
+
+                if (length == value.Length) // Must be all ASCII...
+                {
+                    for (int i = 0; i < length; i++)
+                    {
+                        buffer[i] = (byte)value[i];
+                    }
+
+                    writer.Advance(length);
+                }
+                else
+                {
+                    ReadOnlySpan<char> source = value.AsSpan();
+
+                    int bytesUsed;
+
+                    unsafe
+                    {
+                        fixed (char* sourceChars = &MemoryMarshal.GetReference(source))
+                        fixed (byte* destinationBytes = &MemoryMarshal.GetReference(buffer))
+                        {
+                            bytesUsed = Utf8Encoding.GetBytes(sourceChars, source.Length, destinationBytes, buffer.Length);
+                        }
+                    }
+
+                    writer.Advance(bytesUsed);
+                }
+            }
+            else
+            {
+                // The destination byte array might not be large enough so multiple writes are sometimes required
+                if (encoder == null)
+                {
+                    encoder = Utf8Encoding.GetEncoder();
+                }
+
+                ReadOnlySpan<char> source = value.AsSpan();
+                int written = 0;
+
+                while (true)
+                {
+                    int bytesUsed;
+                    int charsUsed;
+
+                    unsafe
+                    {
+                        fixed (char* sourceChars = &MemoryMarshal.GetReference(source))
+                        fixed (byte* destinationBytes = &MemoryMarshal.GetReference(buffer))
+                        {
+                            encoder.Convert(sourceChars, source.Length, destinationBytes, buffer.Length, false, out charsUsed, out bytesUsed, out _);
+                        }
+                    }
+
+                    source = source.Slice(charsUsed);
+                    written += bytesUsed;
+
+                    writer.Advance(bytesUsed);
+
+                    if (source.Length == 0)
+                    {
+                        break;
+                    }
+
+                    buffer = writer.GetSpan(length - written);
```

Comment:
Maybe ask for more than needed (to potentially avoid having to ask for more later)? Whenever I ask for more output space within the `Utf8JsonWriter`, I ask for 4k: https://github.com/dotnet/corefx/blob/18d9c90e53020e8ab98d7341116e3667e4bc058a/src/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.cs#L793


Although, since GetSpan now returns at least what you asked for, maybe its fine to ask for exactly what you needed.

Reply:
`CodedOutputWriter` is calling `IBufferWriter<byte>` via `BufferWriter` which is caching the span and minimizing the number of calls to the interface.

I may rewrite `CodedOutputWriter` to be more like `Utf8JsonWriter` and party on a local span field for efficiency.